### PR TITLE
Stats collector handle post with retry and graceful shutdown

### LINKED ISSUE #5048: Improve stats posting
Handle posting stats better by using retry. Make sure we post data while we shutdown in order to prevent stats lost.

### PR DESCRIPTION
**Pull request by nopcoder**
Close #5048

- retry on client error while sending metrics (configured to have 2 more tries)
- remove low timeout while posting events.
- sending metrics will not abort on cancellation.
- separate background processing of metrics and one sending metrics.
- shutdown using context cancellation
- first audit check done in the background

### CODE REVIEW

File: pkg/stats/collector.go [hunk 1]
```diff
@@ -104,30 +103,29 @@ func (t *TimeTicker) Tick() <-chan time.Time {
...
-	ctxCancelled     int32
-	done             chan bool
-	runCalled        int32
+	inFlight         sync.WaitGroup
+	closed           int32
 	extended         bool
 	log              logging.Logger
+	wg               sync.WaitGroup
+	cancel           context.CancelFunc
+	sendCh           chan *InputEvent
```
Comment (itaiad200) [inline pkg/stats/collector.go:121]: Seems like we added or replaced many synchronization protocols. Can you document each one for the next developer?

File: pkg/stats/collector.go [hunk 2]
```diff
@@ -173,20 +171,18 @@ func NewBufferedCollector(installationID string, cfg Config, opts ...BufferedCol
...
-		pendingWrites:   sync.WaitGroup{},
-		pendingRequests: sync.WaitGroup{},
-		ctxCancelled:    0,
-		done:            make(chan bool),
-		sendTimeout:     sendTimeout,
+		inFlight:        sync.WaitGroup{},
+		closed:          0,
 		extended:        cfg.Extended,
 		log:             logging.Default(),
+		sendCh:          make(chan *InputEvent, 1), // buffered as like to check if sender is free
```
Comment (itaiad200) [inline pkg/stats/collector.go:185]: Didn't understand your comment.

File: pkg/stats/collector.go [hunk 3]
```diff
@@ -211,154 +207,187 @@ func (s *BufferedCollector) getInstallationID() string {
...
+func (s *BufferedCollector) processSendCh(ctx context.Context) {
+	for event := range s.sendCh {
+		s.sendEventHelper(ctx, event)
+	}
+}
 
-	// wait for http requests to complete
-	s.pendingRequests.Wait()
+func (s *BufferedCollector) sendEventHelper(ctx context.Context, event *InputEvent) {
+	err := s.sender.SendEvent(ctx, event)
```
Comment (itaiad200) [inline pkg/stats/collector.go:358]: pass context.Background to avoid throwing away requests.

File: pkg/stats/collector.go [hunk 4]
```diff
@@ -211,154 +207,187 @@ func (s *BufferedCollector) getInstallationID() string {
...
+	// stop main loop processing
+	s.cancel()
+	s.wg.Wait()
 
-	// drain writes
-	close(s.writes)
-	for w := range s.writes {
+	// drain events
+	ctx := context.Background()
+	s.processSendCh(ctx)
```
Comment (itaiad200) [inline pkg/stats/collector.go:339]: redundant, that function is self draining

File: pkg/stats/collector.go [hunk 5]
```diff
@@ -211,154 +207,187 @@ func (s *BufferedCollector) getInstallationID() string {
...
-	if atomic.LoadInt32(&s.runCalled) == 0 {
-		// nothing to do
+	// mark as closed
+	if !atomic.CompareAndSwapInt32(&s.closed, 0, 1) {
 		return
 	}
 
-	// wait for main loop to exit
-	<-s.done
+	// wait until no more in flight requests
```
Comment (itaiad200) [inline pkg/stats/collector.go:325]: What if there's always more incoming requests?

**Review by itaiad200 — APPROVED** [review]
Hopefully you found it 🤞

**Review by itaiad200 — APPROVED** [review]

File: pkg/stats/collector.go (hunk 4 above)
Comment (nopcoder) [reply pkg/stats/collector.go]: fixed

File: pkg/stats/collector.go (hunk 2 above)
Comment (nopcoder) [reply pkg/stats/collector.go:200]: single item channel to handle one post at a time

File: pkg/stats/collector.go (hunk 3 above)
Comment (nopcoder) [reply pkg/stats/collector.go:377]: fixed

File: pkg/stats/collector.go (hunk 1 above)
Comment (nopcoder) [reply pkg/stats/collector.go]: added comments

File: pkg/stats/collector.go (hunk 5 above)
Comment (nopcoder) [reply pkg/stats/collector.go]: we set the 'stop' flag so no new request will send anything on this channel

**Pull request merged**